### DISCUSSION TITLE
Applying 2 types of BCs on a single surface (surface + nodeset) [PorousFlow]

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#21751] [Q&A Modules: General] [answered]
Hi guys, so I have a surface and a nodeset in a single mesh

The nodeset (blue dots) represents the river on this surface

![image](https://user-images.githubusercontent.com/79562439/182388352-e2c95474-825e-44dc-a2bb-6ab5bc5de416.png)

for the surface I'd like to apply this BC

```
[./drain_TOPOLSRF_Y2008_M01] 
type = PorousFlowPiecewiseLinearSink 
variable = porepressure 
boundary = 'TOPOLSRF_Y2008_M01'
pt_vals = '-1e9 1e9' 
multipliers = '-1e9 1e9' 
flux_function = 1 
v = min
PT_shift = -98100 
use_mobility = true
fluid_phase = 0
[../]
```

and the nodeset I just want to set pwp of those nodes = 0

I know that MOOSE cannot project BC directly into a nodeset (which is very weird), but I know how to work around it (see the thread below)
https://github.com/idaholab/moose/discussions/21209

Now I want both BCs at the same time i.e. pwp of those nodes = 0 & PorousFlowPiecewiseLinearSink on the rest of the surface.

Is there a way to work around this?

Kind regards,
Traiwit

**Comment by Traiwit** [comment]
I see MOOSE can convert node list into side list 

https://mooseframework.inl.gov/source/mesh/MooseMesh.html

`  construct_side_list_from_node_list=true
`
However, it converts all the nodesets into sidesets
is there a way to specifically convert only 1 nodeset into a surface?

**Reply by GiudGiud** [reply]
not currently. We are adding a generator which should be able to do this in its final form here:
https://github.com/idaholab/moose/pull/21196
but currently it still does all sidesets / nodesets.

What you could do is turn the conversions off in MOOSE and do it somewhere else like in Cubit

**Reply by Traiwit** [reply]
Hi @GiudGiud, so `NodeSetToSideSetGenerator` from that PR still does not do what I want, right? I will need to wait then

**Reply by GiudGiud** [reply]
yeah both are still geared towards all sidesets/ all nodesets.

But you should be able to rename part of a nodeset or a sideset. So while all are converted, you may be able to extract the one you care about?

**Comment by GiudGiud** [comment, marked as answer]
PorousFlowPiecewiseLinearSink may be set naturally I think then for `pwp of those nodes = 0` you could look at using a NodalKernel to set the BC instead, maybe with a penalty.

What do you get when trying to set both directly?

**Reply by Traiwit** [reply]
> What do you get when trying to set both directly?

I just realised that https://github.com/idaholab/moose/discussions/21209 wouldn't work, because in that case I specify pwp values to the target nodes and leave pwp the rest to 0. However, in this particular case, I want my target nodes pwp to be 0 and don't assign anything to the rest. (pwp = pore water pressure)

**Reply by GiudGiud** [reply]
do you have a nodeset that only has the nodes you want to set the pwp to 0 ?

**Reply by Traiwit** [reply]
@GiudGiud `NodalKernel` works, at least it starts solving, I will get back to you once I get the results.

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
![image](https://user-images.githubusercontent.com/79562439/183058026-8b737d5e-08ba-4ac8-85bd-7cd3efc02af9.png)

all works fine now, thank you @GiudGiud 

But yeh, I'd rather move `NodalKernel` into `Nodal BC` or we should have this type of object, to us it's like the most basic type of BCs (even more basic than surface BC)

Kind regards,
Traiwit

```
[BCs]
[./water_grad_lTOPOLSRF_WALLS_SRFSET]
type = FunctionDirichletBC
variable = porepressure
boundary = 'TOPOLSRF_WALLS_SRFSET'
function = ini_pp
[../]
[./drain_DRAINSRF_Y1996_M01]
type = PorousFlowPiecewiseLinearSinkBEH
variable = porepressure
boundary = 'DRAINSRF_Y1996_M01'
pt_vals = '0 1e9'
multipliers = '0 1e9'
flux_function = 1
v = min
PT_shift = 0
use_mobility = true
fluid_phase = 0
[../]
[./drain_TOPOLSRF_Y1996_M01]
type = PorousFlowPiecewiseLinearSinkBEH
variable = porepressure
boundary = 'TOPOLSRF_Y1996_M01'
pt_vals = '-1e9 1e9'
multipliers = '-1e9 1e9'
flux_function = 1e-06
v = min
PT_shift = -29430
use_mobility = true
fluid_phase = 0
[../]
[]
[NodalKernels]
[./bc_river]
type = PenaltyDirichletNodalKernel
variable = porepressure
value = 0
boundary = 'TOPO_RIVERSANDLAKES_NSET'
penalty = 1e10
[../]
[]
```

**Reply by GiudGiud** [reply]
I'm glad. If you get difficulties with the penalty you could try a constraint with a scalar variable lagrange multiplier btw

Well nodesets are very alike, whether internal or external to the mesh, and also the action done by nodal kernels is supposed to be local to a single node and not really be too aware of its spatial location. Anyway maybe it's not set in stone but I think that s why we dont have nodal BCs, only nodal kernels.
